storage/postgres/abstract.go

```diff
@@ -86,21 +89,91 @@ func create(ctx context.Context, db pgDB, table string, dto interface{}) (string
 		err = stmt.GetContext(ctx, &lastInsertId, dto)
 		return lastInsertId, checkIntegrityViolation(ctx, checkUniqueViolation(ctx, err))
 	}
-	log.C(ctx).Debugf("Executing query %s", query)
-	_, err := db.NamedExecContext(ctx, query, dto)
-
+	log.C(ctx).Debugf("Executing query %s", sqlQuery)
+	_, err := db.NamedExecContext(ctx, sqlQuery, dto)
 	return lastInsertId, checkIntegrityViolation(ctx, checkUniqueViolation(ctx, err))
 }
 
 func get(ctx context.Context, db getterContext, id string, table string, dto interface{}) error {
-	query := "SELECT * FROM " + table + " WHERE id=$1"
-	log.C(ctx).Debugf("Executing query %s", query)
-	err := db.GetContext(ctx, dto, query, &id)
+	sqlQuery := "SELECT * FROM " + table + " WHERE id=$1"
+	log.C(ctx).Debugf("Executing query %s", sqlQuery)
+	err := db.GetContext(ctx, dto, sqlQuery, &id)
 	return checkSQLNoRows(err)
 }
 
+func listWithLabelsAndCriteria(ctx context.Context, db pgDB, baseEntity interface{}, labelsEntity Labelable, baseTableName string, labelsTableName string, criteria []query.Criterion) (*sqlx.Rows, error) {
+	if err := validateFieldQueryParams(baseEntity, criteria); err != nil {
+		return nil, err
+	}
+	baseQuery := constructBaseQueryForLabelable(labelsEntity, baseTableName, labelsTableName)
+	sqlQuery, queryParams, err := buildQueryWithParams(db, baseQuery, baseTableName, labelsTableName, criteria)
+	if err != nil {
+		return nil, err
+	}
+	log.C(ctx).Debugf("Executing query %s", sqlQuery)
+	return db.QueryxContext(ctx, sqlQuery, queryParams...)
+}
+
+func listByFieldCriteria(ctx context.Context, db pgDB, table string, entity interface{}, criteria ...query.Criterion) error {
+	baseQuery := fmt.Sprintf(`SELECT * FROM %s`, table)
+	sqlQuery, queryParams, err := buildQueryWithParams(db, baseQuery, table, "", criteria)
+	if err != nil {
+		return err
+	}
+	return db.SelectContext(ctx, entity, sqlQuery, queryParams...)
+}
+
+func deleteAllByFieldCriteria(ctx context.Context, extContext sqlx.ExtContext, table string, dto interface{}, criteria ...query.Criterion) error {
+	for _, criterion := range criteria {
+		if criterion.Type != query.FieldQuery {
+			return &query.UnsupportedQuery{Message: "conditional delete is only supported for field queries"}
```

Comment:
Here is one idea to support delete by labelquery and fieldquery and no query - all deletes will use subquery

1. DELETE FROM vis WHERE id IN (SELECT visibility_id FROM vis) - no criterias means to where clause

2. DELETE FROM vis WHERE id IN (SELECT visibility_id FROM vis_labels WHERE key = org_id AND val IN [1,2,3,4]) - labelquery means buildQueryWithParams func builds us a where clause for labels

3. DELETE FROM vis WHERE id IN (SELECT visibility_id FROM vis WHERE field1 = value AND field2 IN [v1,v2,v3,v4] - fieldquery means we use vis table instead of vis_labels in the subquery and we enter the else of buildQueryWithParams func when building the where clause

4.  DELETE FROM vis WHERE id IN (SELECT from 2.) OR id IN (SELECT from 3.) - in case both fieldquery and labelquery - append them with or/union

And another way would be to have a switch with fallthrough in the deleteAllByFieldCriteria and in both cases we would use buildqueryWithParams to build the where clause but in the labelQuery case the where clause would be prepended with WHERE id IN (SELECT id from <table> <whereclause-from-before>) or with AND id IN (SELECT id from <table> <whereclause-from-before>) if we entered the first switch case (the fieldquery one)

So the statement would look like DELETE FROM vis WHERE field=val AND field2 IN [...] AND id IN (SELECT id from <table> <whereclause-from-before>)

Reply:
It would be nice to support this, but I'm not sure if we can't leave it in a new PR